pkg/storage/endpoint/safe_point_v2.go

```diff
@@ -0,0 +1,220 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
+
+import (
+	"encoding/json"
+	"math"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/errs"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+)
+
+// GCSafePointV2 represents the overall safe point for a specific keyspace.
+type GCSafePointV2 struct {
+	KeyspaceID uint32 `json:"keyspace_id"`
+	SafePoint  uint64 `json:"safe_point"`
+}
+
+// ServiceSafePointV2 represents a service's safepoint under a specific keyspace.
+// Services can post service safe point to prevent gc safe point from incrementing.
+type ServiceSafePointV2 struct {
+	KeyspaceID uint32 `json:"keyspace_id"`
+	ServiceID  string `json:"service_id"`
+	ExpiredAt  int64  `json:"expired_at"`
+	SafePoint  uint64 `json:"safe_point"`
+}
+
+// SafePointV2Storage defines the storage operations on safe point v2.
+type SafePointV2Storage interface {
+	LoadGCSafePointV2(keyspaceID uint32) (*GCSafePointV2, error)
+	SaveGCSafePointV2(gcSafePoint *GCSafePointV2) error
+	LoadAllGCSafePoints() ([]*GCSafePointV2, error)
+
+	LoadMinServiceSafePointV2(keyspaceID uint32, now time.Time) (*ServiceSafePointV2, error)
+	LoadServiceSafePointV2(keyspaceID uint32, serviceID string) (*ServiceSafePointV2, error)
+
+	SaveServiceSafePointV2(serviceSafePoint *ServiceSafePointV2) error
+	RemoveServiceSafePointV2(keyspaceID uint32, serviceID string) error
+}
+
+var _ SafePointV2Storage = (*StorageEndpoint)(nil)
+
+// LoadGCSafePointV2 loads gc safe point for the given keyspace.
+func (se *StorageEndpoint) LoadGCSafePointV2(keyspaceID uint32) (*GCSafePointV2, error) {
+	key := GCSafePointV2Path(keyspaceID)
+	value, err := se.Load(key)
+	if err != nil {
+		return nil, err
+	}
+	// GC Safe Point does not exist for the given keyspace
+	if value == "" {
+		return &GCSafePointV2{
+			KeyspaceID: keyspaceID,
+			SafePoint:  0,
+		}, nil
+	}
+	gcSafePoint := &GCSafePointV2{}
+	if err = json.Unmarshal([]byte(value), gcSafePoint); err != nil {
+		return nil, errs.ErrJSONUnmarshal.Wrap(err).GenWithStackByCause()
+	}
+	return gcSafePoint, nil
+}
+
+// SaveGCSafePointV2 saves gc safe point for the given keyspace.
+func (se *StorageEndpoint) SaveGCSafePointV2(gcSafePoint *GCSafePointV2) error {
+	key := GCSafePointV2Path(gcSafePoint.KeyspaceID)
+	value, err := json.Marshal(gcSafePoint)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	return se.Save(key, string(value))
+}
+
+// LoadAllGCSafePoints returns gc safe point for all keyspaces
+func (se *StorageEndpoint) LoadAllGCSafePoints() ([]*GCSafePointV2, error) {
+	prefix := GCSafePointV2Prefix()
+	prefixEnd := clientv3.GetPrefixRangeEnd(prefix)
+	_, values, err := se.LoadRange(prefix, prefixEnd, 0)
+	if err != nil {
+		return nil, err
+	}
+	gcSafePoints := make([]*GCSafePointV2, 0, len(values))
+	for _, value := range values {
```

Comment:
Is it possible that any value is an empty string? If so, json.Unmarshal would fail.